setup.py

```diff
@@ -519,21 +540,229 @@ def add_dir_name(dir_name, filenames):
+
+            cmake_cmd.append(
+                '-DTICKLIB_{}={}'.format(mod.ext_name.upper(),
+                                            full_path))
+
+        os.makedirs(os.path.join(self.cpp_build_dir, 'cpptest'), exist_ok=True)
+        subprocess.check_call(cmake_cmd, cwd=self.cpp_build_dir)
+
+        make_cmd = ['make', 'all', '-j{}'.format(self.build_jobs)]
+        subprocess.check_call(make_cmd, cwd=self.cpp_build_dir)
+
+
+class RunCPPLint(TickCommand):
+    description = 'run cpplint on tick C++ source files'
+
+    CPPLINT_DIRS = [
+        'tick/base/array/src',
+        'tick/base/array/tests/src',
+        'tick/base/array_test/src',
+        'tick/base/src',
+        'tick/base/src/math',
+        'tick/base/src/parallel',
+        'tick/inference/src',
+        'tick/optim/model/src',
+        'tick/optim/solver/src',
+        'tick/optim/prox/src',
+        'tick/random/src',
+        'tick/simulation/src',
+    ]
+
+    def run(self):
+        try:
+            import cpplint as cl
+
+            cl_state = cl._cpplint_state
+            error_count = 0
+
+            for dir in self.CPPLINT_DIRS:
+                print("Processing {}".format(dir))
+
+                cl_state.ResetErrorCounts()
+                filenames = list(pathlib.Path(dir).glob('**/*.h')) + \
+                            list(pathlib.Path(dir).glob('**/*.cpp'))
+
+                for filename in filenames:
+                    cl.ProcessFile(str(filename), cl_state.verbose_level)
+                cl_state.PrintErrorCounts()
+
+                error_count += cl_state.error_count
+                print('')
+
+            if error_count > 0:
+                raise RuntimeError("Codestyle check by cpplint failed")
+
+        except ImportError:
+            warnings.warn("Stylecheck by cpplint failed because cpplint "
+                          "is not installed as a Python module")
+
+
+class RunPyLint(TickCommand):
+    # We need to define if and how we run pylint
+
+    description = 'run tick PyLint codestyle check'
+    start_dir = '.'
+
+
+class RunPyTests(TickCommand):
+    description = 'run tick Python tests'
+    start_dir = '.'
+
+    user_options = [
+        ('start-dir=', 's',
+         'directory to start looking for Python tests (e.g. tick/simulation)'),
+    ]
+
+    def initialize_options(self):
+        """Set default values for options."""
+        self.start_dir = '.'
+
+    def run(self):
+        loader = unittest.TestLoader()
+        alltests = loader.discover(self.start_dir, pattern="*_test.py")
+
+        unittest.TextTestRunner(verbosity=2).run(alltests)
+
+
+class RunTestSuites(TickCommand):
+    description = 'run tick Python and C++ tests'
+
+    def run(self):
+        self.run_command('cpptest')
+        self.run_command('pytest')
+
+
+class CleanTick(clean):
+    description = 'cleans all generated and built files'
+
+    def run(self):
```

Comment:
should'nt we add a `time.sleep(2)` here such that user has 2 seconds to cancel full clean?